subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/topology_aware_scheduler.go

```diff
@@ -0,0 +1,284 @@
+// MIT License
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE
+
+package algorithm
+
+import (
+	"fmt"
+	"sort"
+)
+
+// topologyAwareScheduler can schedule a set of pods on a cluster view.
+type topologyAwareScheduler struct {
+	clusterView              CellList  // a list of node- (or lower-)level cells
+	podGpuNumbers            []int32   // GPU number of each pod to schedule
+	currentSolutionIndices   []int32   // indices of currently picked nodes (or GPUs)
+	currentLCA               CellList  // current LCA (lowest common ancestor) of currently picked GPUs
+	lowestLCASolution        CellList  // GPUs with the lowest LCA
+	lowestLCASolutionIndices []int32   // Indices of the GPUs with the lowest LCA
+	lowestLCALevel           CellLevel // level of the lowest LCA
+	// pack pods cross different priorities, or inside each priority. the former is for opportunistic
```

Comment:
>// pack pods cross different priorities, or inside each priority. the former is for opportunistic [](start = 1, length = 97)

If so, why opportunistic's crossPriorityPack = false

Reply:
I wrote it wrong... They are in the wrong order. I've corrected and refined the comment.

---
In reply to: [332856608](https://github.com/microsoft/pai/pull/3684#discussion_r332856608) [](ancestors = 332856608)